jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/VersionChecker.java

```diff
@@ -0,0 +1,143 @@
+ * Note that Gradle's {@code org.gradle.util.GradleVersion} class does not support
+ * major-version-only versions.
+ */
+public class VersionChecker<V extends Comparable<? super V>> {
+  /** The expected version representation. */
+  private static final String VERSION_PATTERN = "\\d+(\\.\\d+(\\.\\d+)?)?";
+
+  // Helper functions to avoid the cognitive burden of {@link Comparable#compareTo()}
+
+  /** Return {@code true} if {@code a} is less than {@code b}. */
+  @VisibleForTesting
+  static <T extends Comparable<? super T>> boolean LT(T a, T b) {
+    return a.compareTo(b) < 0;
+  }
+
+  /** Return {@code true} if {@code a} is less than or equal to {@code b}. */
+  @VisibleForTesting
+  static <T extends Comparable<? super T>> boolean LE(T a, T b) {
+    return a.compareTo(b) <= 0;
+  }
+
+  /** Return {@code true} if {@code a} is greater than {@code b}. */
+  @VisibleForTesting
+  static <T extends Comparable<? super T>> boolean GT(T a, T b) {
+    return a.compareTo(b) > 0;
+  }
+
+  /** Return {@code true} if {@code a} is greater than or equal to {@code b}. */
+  @VisibleForTesting
+  static <T extends Comparable<? super T>> boolean GE(T a, T b) {
+    return a.compareTo(b) >= 0;
+  }
+
+  /** Responsible for converting a string representation to a comparable version representation. */
+  private Function<String, V> converter;
+
+  public VersionChecker(Function<String, V> converter) {
+    this.converter = converter;
+  }
+
+  /**
+   * Return {@code true} if {@code actualVersion} is contained within the version range represented
+   * by {@code acceptableVersionRange}.
+   *
+   * @param acceptableVersionRange the encoded version range
+   * @param actualVersion the version to be compared
+   * @return true if the version is acceptable
+   * @throws IllegalArgumentException if the version could not be parsed
+   */
+  @VisibleForTesting
+  public boolean compatibleVersion(String acceptableVersionRange, String actualVersion) {
+    V pluginVersion = parseVersion(actualVersion);
+
+    // Treat a single version "1.4" as a left bound, equivalent to "[1.4,)"
+    if (acceptableVersionRange.matches(VERSION_PATTERN)) {
+      return GE(pluginVersion, parseVersion(acceptableVersionRange));
+    }
+
+    // Otherwise ensure it is a version range with bounds
+    Preconditions.checkArgument(
+        acceptableVersionRange.matches(
+            "[\\[(](" + VERSION_PATTERN + ")?,(" + VERSION_PATTERN + ")?[\\])]"),
+        "invalid version range");
+    BiPredicate<V, V> leftComparator =
+        acceptableVersionRange.startsWith("[") ? VersionChecker::GE : VersionChecker::GT;
+    BiPredicate<V, V> rightComparator =
+        acceptableVersionRange.endsWith("]") ? VersionChecker::LE : VersionChecker::LT;
+    // extract the two version specs
+    String[] range =
+        acceptableVersionRange.substring(1, acceptableVersionRange.length() - 1).split(",", -1);
```

Comment:
Is it not possible to get the versions directly from the regex group?

Reply:
Good point.